### PR TITLE
Value validation

### PR DESCRIPTION
**Pull request by mateuszsip**
Changes value validation implementation including support for `@Enum` and `@Required`.

This was implemented 10 days ago so I have to do refactoring based on your review:
- [x] all assertions should be called statically
- [x] data providers should be suffixed with "Provider"
- [x] all tests should be final

But we can discuss design at this point.

### CODE REVIEW

File: lib/Doctrine/Annotations/Metadata/AnnotationMetadata.php [hunk 1]
```diff
@@ -79,4 +81,40 @@ public function getDefaultProperty() : ?PropertyMetadata
...
+            if (! $target->annotation()) {
+                throw InvalidTarget::annotation($this);
+            }
+
+            return;
+        }
+
+        $subject = $scope->getSubject();
+
+        if ($subject instanceof ReflectionClass && ! $target->class()) {
```
Comment (mateuszsip) [inline lib/Doctrine/Annotations/Metadata/AnnotationMetadata.php]: Maybe target should take care of that?

File: lib/Doctrine/Annotations/Metadata/Constraint/Constraint.php [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+interface Constraint
+{
+    /**
+     * @param mixed $value
+     *
+     * @return true
+     *
+     * @throws ConstraintNotFulfilled
+     */
+    public function validate($value) : bool;
```
Comment (Majkl578) [inline lib/Doctrine/Annotations/Metadata/Constraint/Constraint.php]: This should be void - either throws or does nothing.

File: lib/Doctrine/Annotations/Metadata/Constraint/EnumConstraint.php [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+        $this->allowedValues = $allowedValues;
+    }
+
+    /**
+     * @param mixed $value
+     */
+    public function validate($value) : bool
+    {
+        if (! in_array($value, $this->allowedValues)) {
+            throw InvalidValue::new($this->allowedValues, $value);
```
Comment (Majkl578) [inline lib/Doctrine/Annotations/Metadata/Constraint/EnumConstraint.php]: These conditions should be inverted so happy path returns here and unhappy path throws below (once `return true` is gone).

File: lib/Doctrine/Annotations/Metadata/Constraint/InvalidValue.php [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+     * @param mixed   $value
+     */
+    public static function new(array $allowedValues, $value) : self
+    {
+        // TODO: Describe values
+        return new self(
+            sprintf(
+                'Invalid value "%s" for allowed values: "%s".',
+                gettype($value),
+                implode(',', $allowedValues)
```
Comment (Majkl578) [inline lib/Doctrine/Annotations/Metadata/Constraint/InvalidValue.php]: `$allowedValues` is `mixed`, but `implode()` assumes strings.

File: lib/Doctrine/Annotations/Metadata/InvalidTarget.php [hunk 5]
```diff
@@ -2,9 +2,8 @@
 
 declare(strict_types=1);
 
-namespace Doctrine\Annotations\Assembler\Validator\Exception;
+namespace Doctrine\Annotations\Metadata;
```
Comment (Majkl578) [inline lib/Doctrine/Annotations/Metadata/InvalidTarget.php]: Should be in Exception sub-namespace.

File: lib/Doctrine/Annotations/Metadata/PropertyMetadata.php [hunk 6]
```diff
@@ -4,38 +4,52 @@
...
+use Doctrine\Annotations\Metadata\Constraint\Constraint;
+use Doctrine\Annotations\Metadata\Constraint\ConstraintNotFulfilled;
 
 final class PropertyMetadata
 {
     /** @var string */
     private $name;
 
-    /** @var Type */
-    private $type;
```
Comment (Majkl578) [inline lib/Doctrine/Annotations/Metadata/PropertyMetadata.php]: As discussed, Constraint should not replace `$type` here but build on top of it instead.

**Review by Majkl578 — COMMENTED** [review]
comments for 08d35dbdb8302c95cee0b6d8d7bc79947a6c07ef

File: lib/Doctrine/Annotations/Metadata/Constraint/Constraint.php (hunk 2 above)
Comment (Majkl578) [reply lib/Doctrine/Annotations/Metadata/Constraint/Constraint.php]: Resolved in next commit.

File: lib/Doctrine/Annotations/Metadata/Constraint/InvalidValue.php (hunk 4 above)
Comment (mateuszsip) [reply lib/Doctrine/Annotations/Metadata/Constraint/InvalidValue.php]: Looks like my TODO is about it. 
Should we describe values or remove them from message?
Do you know a lib that does this?

**Comment by mateuszsip** [comment]
@Majkl578 one thing is still missing (related to InvalidValue), but imo you can check latest commits

scrutinizer fails look like false positive

File: lib/Doctrine/Annotations/Assembler/Validator/TargetValidator.php [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+
+        if ($subject instanceof ReflectionProperty && ! $target->property()) {
+            throw InvalidTarget::property($metadata);
+        }
+
+        if ($subject instanceof ReflectionMethod && ! $target->method()) {
+            throw InvalidTarget::method($metadata);
+        }
+
+        // TODO validate annotation target
```
Comment (Majkl578) [inline lib/Doctrine/Annotations/Assembler/Validator/TargetValidator.php]: This is handled on line 27 - you can move that block here and change it to early-exit for simplification.

**Pull request closed without merging**